File: cplusplus/024_Swap_Nodes_in_Pairs.cpp
/*
Solution 1
*/
class Solution {
public:
  ListNode* swapPairs(ListNode* head) {
    if(!head || !head->next) return head;
    ListNode* next = head->next;
    head->next = swapPairs(head->next->next);
    next->next = head;
    return next;
  }
};

/*
Solution 2
*/
class Solution {
public:
  ListNode* swapPairs(ListNode* head) {
    ListNode dummy(0),*pre(&dummy),*left(nullptr),*right(nullptr);
    dummy.next = head;
    while(pre && pre->next && pre->next->next){
      left = pre->next;
      right = left->next;

      pre->next = right;
      left->next = right->next;
      right->next = left;
      pre = left;
    }
    return dummy.next;
  }
};

File: java/024_Swap_Nodes_in_Pairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode swapPairs(ListNode head) {
        ListNode dummy = new ListNode(0), pre = dummy;
        dummy.next = head;
        
        while (pre.next != null && pre.next.next != null) {
            ListNode left = pre.next, right = left.next;
            pre.next = right;
            left.next = right.next;
            right.next = left;
            pre = left;
        }
        
        return dummy.next;
    }
}
